cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -384,20 +384,69 @@ func generateInitramfsMounts() error {
 }
 
 func unlockIfEncrypted(name string) (string, error) {
-	// TODO:UC20: will need to unseal key to unlock LUKS here
 	device := filepath.Join("/dev/disk/by-label", name)
-	keyfile := filepath.Join(boot.InitramfsUbuntuBootDir, name+".keyfile.unsealed")
-	if osutil.FileExists(keyfile) {
+	encdev := device + "-enc"
+	if osutil.FileExists(encdev) {
 		// TODO:UC20: snap-bootstrap should validate that <name>-enc is what
 		//            we expect (and not e.g. an external disk), and also that
 		//            <name> is from <name>-enc and not an unencrypted partition
 		//            with the same name (LP #1863886)
-		cmd := exec.Command("/usr/lib/systemd/systemd-cryptsetup", "attach", name, device+"-enc", keyfile)
-		cmd.Env = os.Environ()
-		cmd.Env = append(cmd.Env, "SYSTEMD_LOG_TARGET=console")
-		if output, err := cmd.CombinedOutput(); err != nil {
-			return "", osutil.OutputErr(output, err)
+		sealedKeyPath := filepath.Join(boot.InitramfsEncryptionKeyDir, name+".keyfile.sealed")
+		if err := unlockEncryptedPartition(name, encdev, sealedKeyPath, "", ""); err != nil {
+			return device, fmt.Errorf("cannot unlock %s: %v", name, err)
 		}
 	}
 	return device, nil
 }
+
+var (
+	secbootConnectToDefaultTPM            = secboot.ConnectToDefaultTPM
+	secbootSecureConnectToDefaultTPM      = secboot.SecureConnectToDefaultTPM
+	secbootActivateVolumeWithTPMSealedKey = secboot.ActivateVolumeWithTPMSealedKey
+)
+
+func secureConnectToTPM(ekcfile string) (*secboot.TPMConnection, error) {
+	ekCertReader, err := os.Open(ekcfile)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open endorsement key certificate file: %v", err)
+	}
+	defer ekCertReader.Close()
+
+	return secbootSecureConnectToDefaultTPM(ekCertReader, nil)
+}
+
+func insecureConnectToTPM(ekcfile string) (*secboot.TPMConnection, error) {
+	return secbootConnectToDefaultTPM()
+}
+
+// unlockEncryptedPartition unseals the keyfile and opens an encrypted device.
+func unlockEncryptedPartition(name, device, keyfile, ekcfile, pinfile string) error {
+
+	// TODO:UC20: use secureConnectToTPM
+	tpm, err := insecureConnectToTPM(ekcfile)
```

Comment:
can we just use secureConnectToTPM in this PR yet? I see that the relevant function is already defined? If not, could we remove that function definition from here?

Reply:
In order to establish a secure connection to the TPM we need to download the EK certificate. This will be done before the final release but we need to establish how and when we'll download it (it must be after network is set up) and what we'll do if the download fails.